### ISSUE TITLE
Compilcated debugging of sending commands via WS

### ISSUE THREAD

**Issue by akirilyuk** [jambonz/jambonz-feature-server#85]
Hi,

when sending commands such as redirect via the websocket connection instead of using webhooks as app, it's is kind of complicated to debug them. Especially when queuing commands, its hard to debug them. It would be great to have a feedback look so when we are sending commands to  feature server it  would provide us some feedback.

My proposal would be the following:

Each command send should contain a id. So for example if we send a list of commands via the redirect command, the payload would look like this:

```js
{
				type: "command",
				command: "redirect",
				queueCommand: true,
				data: [
					{
						gather: {
							input: ["speech"],
							listenDuringPrompt: false,
							bargein: false,
							minBargeinWordCount: 0,
							dtmfBargein: false,
							minDigits: 1,
							interDigitTimeout: 0,
							recognizer: {
								vendor: "default",
								language: "default",
								vad: { enable: true, mode: 0 }
							},
							actionHook: "voice",
							timeout: 10
						},
                                               verbid: "53edc152-3221-475b-8d61-6da7ea3a668c"
					}
				]
			}
```
Then if its a redirect without quuing commands. Once the command was successfully executed, we get a confirmation back on the ws with the following

```js
{
  verbid :  "53edc152-3221-475b-8d61-6da7ea3a668c",
  status:  "SUCCESSS" | "ABORTED" | "ERROR",
  error: "in case of error send us the error object here (ideallyjambonz should have an own one with some general error codes)"
  metrics: "we could also think of adding some metrics here such as receivedOnTime, executionTime, queueTime," etc
}
```
This would really help debugging and fixing issues faster and also identify wrong send verb configurations without the need of going into the feature-server logs and debugging the error messages there.

What do you think about this idea?

I am happy to provide a PR with a first implementation iteration if you think this is something beneficial for the feature server

**Comment by davehorton** [comment]
I do like the idea.  I have a few thoughts/opinions to contribute:

- I'd prefer simply to name the new property "id".  It can be any string value assigned by (and meaningful to) the sending app
- I'd prefer to name the status notification "verb:status" (consistent with "verb:hook").  verb:status notifications will automatically be sent if (and only if) an "id" property is included in the command
- I think verb:status notifications should be sent both when a verb starts executing as well as when it completes executing. 
- I agree any relevant error information can be conveyed with the completion status of the verb
- I don't think metrics should be part of this, as it should be handled separately via the currently-implemented metrics capability (statsd), or via opentelemetry tracing, which I am currently implementing as well.  The opentelemetry spans created for each verb can include a tag with the id value, if provided

With those changes and a few others your example would be:

```json
{
	"type": "command",
	"command": "redirect",
	"id": "53edc152-3221-475b-8d61-6da7e13a668c",
	"queueCommand": true,
	"data": [{
		"gather": {
			"id": "38f152-8958-475b-8d61-6da7ea2376ed4",
			"input": ["speech"],
			"etc..": "etc.."
		}
	}]
}
```
and then when it executes:
```json
{
	"type": "verb:status",
	"command":  "53edc152-3221-475b-8d61-6da7e13a668c",
	"verb": "38f152-8958-475b-8d61-6da7ea2376ed4",
	"name": "gather"
	"status": "start"
}
```
later..
```json
{
	"type": "verb:status",
	"command":  "53edc152-3221-475b-8d61-6da7e13a668c",
	"verb": "38f152-8958-475b-8d61-6da7ea2376ed4",
	"name": "gather"
	"status": "complete",
	"error": "no speech credentials provisioned"
}
```

I would prefer to do the initial implementation of this myself.

Let me know your thoughts

**Comment by davehorton** [comment]
Also, this behavior would be specific to a "redirect" command

**Comment by akirilyuk** [comment]
Hi Dave, sending the details via verb:status so its consistent with verb:hook sounds reasonable to me. I am OK with the other proposed suggestions and thought. Love to see the the first draft in action.

**Comment by davehorton** [comment]
In the initial spec (https://docs.google.com/document/d/1jtPg_w59PjtLPnYpDnegkUlPiTtHCX270YsPGfYgOcg/edit?usp=sharing) I had provided for a "jambonz:error" message from jambonz -> app if a fatal error of some kind occurs.  I think this would suffice for the scenario you describe where a command is invalid or can not be executed for some reason

**Comment by davehorton** [comment]
no, I am currently not sending any jambonz:error messages.  Its part of the spec though, and I intend to add it as part of this work

**Comment by akirilyuk** [comment]
>I would rather not track verbs that are never executed at all, because the app submitted a new slate of verbs. I would only want to track verbs that actually execute

It would still make things easier to track which command was executed and which was actually skipped execution because we replaced them with a new set of verbs. We could also implement the tracking on the application side, however this would be more complex than just getting another state which would be aborted or whatever from jambonz.
